src/lib.rs

```diff
@@ -502,15 +499,22 @@ pub(crate) unsafe fn rav1d_parse_sequence_header(
     let mut c: *mut Rav1dContext = 0 as *mut Rav1dContext;
     rav1d_open(&mut c, &s)?;
     || -> Rav1dResult<Rav1dSequenceHeader> {
-        if !ptr.is_null() {
-            rav1d_data_wrap_internal(&mut buf, ptr, sz, Some(dummy_free), 0 as *mut c_void)?;
-        }
-
-        while buf.sz > 0 {
-            let len = rav1d_parse_obus(&mut *c, &mut buf, true)?;
-            assert!(len <= buf.sz);
-            buf.sz -= len;
-            buf.data = buf.data.add(len);
+        let Rav1dData {
+            mut data,
+            m: mut props,
+        } = match NonNull::new(ptr.cast_mut()) {
+            None => Default::default(),
+            Some(ptr) => Rav1dData::wrap(
+                slice::from_raw_parts(ptr.as_ptr(), sz).into(),
+                Some(dummy_free),
+                ptr::null_mut(),
+            )?,
+        };
+        if let Some(data) = &mut data {
+            while !data.is_empty() {
```

Comment:
~~Am I correct in reading this as a deref to the inner slice and calling is_empty() on that? It's hard to figure out types here. If so, how is that ever possible? We don't change the slice size, just the stable_ptr field, right?~~

Ok, this is pretty confusing when we're dealing with slices. Probably worth rethinking some of this or making it more explicit. I see now that we are effectively keeping a slice in the stable_ref and that's what is deref'd, not the actual underlying boxed slice. That's kind of what we want, but it isn't obvious that's what will happen.

Reply:
What about making RavidData not a `type` but instead a newtype with a narrower interface? `RavidData` is apparently a ref-counted slice view, which isn't exactly a type Rust already has, so we should be explicit what that type is.

Reply:
Wdym exactly?  It's already a `struct` right now.

Reply:
I must have been looking at a different PR, but the point about all the refs and derefs being confusing is still here. It's confusing to have a type that derefs to an inner type except when it's a slice, when it behaves more like a view into that thing. I think I put a comment on the CArc PR that explains that better.

Reply:
I tried to separate the two, but it proved fairly difficult and very unwieldy.  Can I just improve the docs (I might've added some stuff by now)?

Reply:
@rinon Checking in with you: Is this something to block the PR on? The current version of things seems fairly understandable to me, I'm not sure what has changed between when you last commented and now.

Reply:
@rinon, does this look okay now?